stat/distmat/permutation_test.go

```diff
@@ -0,0 +1,63 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package distmat
+
+import (
+	"math/cmplx"
+	"testing"
+
+	"golang.org/x/exp/rand"
+
+	"gonum.org/v1/gonum/floats"
+	"gonum.org/v1/gonum/mat"
+)
+
+func TestUniformPermutation(t *testing.T) {
+	up := NewUniformPermutation(rand.NewSource(1))
+	for _, n := range []int{10, 32, 64, 100} {
+		m := up.Matrix(n)
+		if m == nil {
+			t.Error("Matrix failed")
+		}
+		if !confirmEigenvaluesAreRootsOfUnity(m) {
+			t.Error("eigenvalue not a root of unity")
+		}
+		if !confirmEachColumnOneNZ(m) {
+			t.Error("didnt get back a permutation matrix")
+		}
+	}
+
+}
+
+func confirmEigenvaluesAreRootsOfUnity(m mat.Matrix) bool {
+	n, _ := m.Dims()
+	var e mat.Eigen
+	e.Factorize(m, mat.EigenLeft)
+	values := make([]complex128, n)
+	e.Values(values)
+	for _, v := range values {
+		a := cmplx.Abs(v)
+		if !floats.EqualWithinAbs(a, 1.0, 1e-12) {
+			return false
+		}
+	}
+	return true
+}
+
+func confirmEachColumnOneNZ(m mat.Matrix) bool {
```

Comment:
This function could also test that m is square and that each column has a 1 in a different row and 0s everywhere else. Then the eigenvalue-based test (which is IMHO quite expensive) would not be necessary.

Reply:
now directly checks for 1 1, n-1 0s and nothing else in each row and column.

Reply:
Thanks!